### ISSUE TITLE
possible watch service leak for DLQ when pipelines crash

### ISSUE THREAD

**Issue by bczifra** [elastic/logstash#12807]
## Bug
If the dead letter queue is not available, the dead letter queue input plugin to not be able to start. This then causes a pipeline reload, which may cause a WatchService leak. 

Logstash logs recorded:
```
[2021-03-22T17:24:59,243][ERROR][logstash.javapipeline    ][nonprod] Pipeline error {:pipeline_id=>"nonprod", :exception=>java.nio.file.NoSuchFileException: /var/lib/logstash/dead_letter_queue/main, :backtrace=>["sun.nio.fs.UnixException.translateToIOException(sun/nio/fs/UnixException.java:92)", "sun.nio.fs.UnixException.asIOException(sun/nio/fs/UnixException.java:120)", "sun.nio.fs.LinuxWatchService$Poller.implRegister(sun/nio/fs/LinuxWatchService.java:248)", "sun.nio.fs.AbstractPoller.processRequests(sun/nio/fs/AbstractPoller.java:265)", "sun.nio.fs.LinuxWatchService$Poller.run(sun/nio/fs/LinuxWatchService.java:366)", "java.lang.Thread.run(java/lang/Thread.java:834)"], "pipeline.sources"=>["/etc/logstash/nonprod.conf.d/100-apex-nonprod-input-collectors.conf", "/etc/logstash/nonprod.conf.d/200-apex-nonprod-id.conf", "/etc/logstash/nonprod.conf.d/220-apex-misc-filter-mcafee.conf", "/etc/logstash/nonprod.conf.d/220-apex-misc-filter-mimecast.conf", "/etc/logstash/nonprod.conf.d/230-apex-nonprod-filter-crowdstrike.conf", "/etc/logstash/nonprod.conf.d/230-apex-nonprod-filter-zscaler.conf", "/etc/logstash/nonprod.conf.d/290-apex-all-filter-postprocessing.conf", "/etc/logstash/nonprod.conf.d/295-all-apex-filter-related.conf", "/etc/logstash/nonprod.conf.d/300-apex-all-output-nonprod.conf"], :thread=>"#<Thread:0x1a6f233d run>"}
```
Note the `java.nio.file.NoSuchFileException: /var/lib/logstash/dead_letter_queue/main` bit.

This got logged 128 times, the same as the value of `fs.inotify.max_user_instances` on the host. Then, the error message changed to:
```

[2021-03-22T23:59:44,901][ERROR][logstash.javapipeline    ][nonprod] Pipeline error {:pipeline_id=>"nonprod", :exception=>java.io.IOException: User limit of inotify instances reached or too many open files, :backtrace=>["sun.nio.fs.LinuxWatchService.<init>(sun/nio/fs/LinuxWatchService.java:64)", "sun.nio.fs.LinuxFileSystem.newWatchService(sun/nio/fs/LinuxFileSystem.java:47)", "org.logstash.common.io.DeadLetterQueueReader.<init>(org/logstash/common/io/DeadLetterQueueReader.java:74)", "org.logstash.input.DeadLetterQueueInputPlugin.<init>(org/logstash/input/DeadLetterQueueInputPlugin.java:48)", "jdk.internal.reflect.GeneratedConstructorAccessor141.newInstance(jdk/internal/reflect/GeneratedConstructorAccessor141)", "jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(jdk/internal/reflect/DelegatingConstructorAccessorImpl.java:45)", "java.lang.reflect.Constructor.newInstance(java/lang/reflect/Constructor.java:490)", "org.jruby.javasupport.JavaConstructor.newInstanceDirect(org/jruby/javasupport/JavaConstructor.java:253)", "org.jruby.RubyClass.newInstance(org/jruby/RubyClass.java:939)", "org.jruby.RubyClass$INVOKER$i$newInstance.call(org/jruby/RubyClass$INVOKER$i$newInstance.gen)", "usr.share.logstash.vendor.bundle.jruby.$2_dot_5_dot_0.gems.logstash_minus_input_minus_dead_letter_queue_minus_1_dot_1_dot_5.lib.logstash.inputs.dead_letter_queue.register(/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/logstash-input-dead_letter_queue-1.1.5/lib/logstash/inputs/dead_letter_queue.rb:54)", "usr.share.logstash.vendor.bundle.jruby.$2_dot_5_dot_0.gems.logstash_minus_input_minus_dead_letter_queue_minus_1_dot_1_dot_5.lib.logstash.inputs.dead_letter_queue.RUBY$method$register$0$__VARARGS__(usr/share/logstash/vendor/bundle/jruby/$2_dot_5_dot_0/gems/logstash_minus_input_minus_dead_letter_queue_minus_1_dot_1_dot_5/lib/logstash/inputs//usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/logstash-input-dead_letter_queue-1.1.5/lib/logstash/inputs/dead_letter_queue.rb)", "usr.share.logstash.logstash_minus_core.lib.logstash.java_pipeline.register_plugins(/usr/share/logstash/logstash-core/lib/logstash/java_pipeline.rb:228)", "org.jruby.RubyArray.each(org/jruby/RubyArray.java:1809)", "usr.share.logstash.logstash_minus_core.lib.logstash.java_pipeline.register_plugins(/usr/share/logstash/logstash-core/lib/logstash/java_pipeline.rb:227)", "usr.share.logstash.logstash_minus_core.lib.logstash.java_pipeline.start_inputs(/usr/share/logstash/logstash-core/lib/logstash/java_pipeline.rb:386)", "usr.share.logstash.logstash_minus_core.lib.logstash.java_pipeline.RUBY$method$start_inputs$0$__VARARGS__(usr/share/logstash/logstash_minus_core/lib/logstash//usr/share/logstash/logstash-core/lib/logstash/java_pipeline.rb)", "usr.share.logstash.logstash_minus_core.lib.logstash.java_pipeline.start_workers(/usr/share/logstash/logstash-core/lib/logstash/java_pipeline.rb:311)", "usr.share.logstash.logstash_minus_core.lib.logstash.java_pipeline.RUBY$method$start_workers$0$__VARARGS__(usr/share/logstash/logstash_minus_core/lib/logstash//usr/share/logstash/logstash-core/lib/logstash/java_pipeline.rb)", "usr.share.logstash.logstash_minus_core.lib.logstash.java_pipeline.run(/usr/share/logstash/logstash-core/lib/logstash/java_pipeline.rb:185)", "usr.share.logstash.logstash_minus_core.lib.logstash.java_pipeline.RUBY$method$run$0$__VARARGS__(usr/share/logstash/logstash_minus_core/lib/logstash//usr/share/logstash/logstash-core/lib/logstash/java_pipeline.rb)", "usr.share.logstash.logstash_minus_core.lib.logstash.java_pipeline.start(/usr/share/logstash/logstash-core/lib/logstash/java_pipeline.rb:137)", "org.jruby.RubyProc.call(org/jruby/RubyProc.java:318)", "java.lang.Thread.run(java/lang/Thread.java:834)"], "pipeline.sources"=>["/etc/logstash/nonprod.conf.d/100-apex-nonprod-input-collectors.conf", "/etc/logstash/nonprod.conf.d/200-apex-nonprod-id.conf", "/etc/logstash/nonprod.conf.d/220-apex-misc-filter-mcafee.conf", "/etc/logstash/nonprod.conf.d/220-apex-misc-filter-mimecast.conf", "/etc/logstash/nonprod.conf.d/230-apex-nonprod-filter-crowdstrike.conf", "/etc/logstash/nonprod.conf.d/230-apex-nonprod-filter-zscaler.conf", "/etc/logstash/nonprod.conf.d/290-apex-all-filter-postprocessing.conf", "/etc/logstash/nonprod.conf.d/295-all-apex-filter-related.conf", "/etc/logstash/nonprod.conf.d/300-apex-all-output-nonprod.conf"], :thread=>"#<Thread:0x1f1e1f02 run>"}
```
where the message notes: `User limit of inotify instances reached or too many open files`.

It appears that for some reason when Logstash was restarted at:
```
2021-03-22T17:15:15,713][INFO ][logstash.runner          ] Starting Logstash {"logstash.version"=>"7.11.1", "jruby.version"=>"jruby 9.2.13.0 (2.5.7) 2020-08-03 9a89c94bcc OpenJDK 64-Bit Server VM 11.0.8+10 on 11.0.8+10 +indy +jit 
```
The folder containing the dead letter queue was not available, causing the dead letter queue input plugin to not be able to start, causing a pipeline reload for that particular pipeline. The DeadLetterQueueReader uses a single java WatchService to watch the dead letter queue folder, which under the hood uses inotify. Once the pipeline had failed to reload 128 times, the attempt to create a new WatchService failed with the error seen subsequently.

This has only occurred once, no reproduction is available. It's also unknown why the DLQ was missing.

**Comment by kesslerm** [comment]
One way to reproduce the startup failure is to have a separate DLQ and processing pipeline. If the DLQ pipeline starts up much more quickly than the main pipeline, it fails to initialise because the `/var/lib/logstash/dead_letter_queue/<pipeline_name>` directory has not yet been created. Maybe the DLQ input plugin needs to either wait for successful creation of expected directories, or create the directory itself.

**Comment by kesslerm** [comment]
Sending SIGHUP to the Logstash process after all other pipelines have been created allows the DLQ input to start correctly.

**Comment by kares** [comment]
the watch service leak is resolved with DLQ input 1.1.11: https://github.com/logstash-plugins/logstash-input-dead_letter_queue/pull/35
it isn't part of any released LS yet but should land in >= 8.1.1 (and also >= 7.17.2), users can also update the plugin
